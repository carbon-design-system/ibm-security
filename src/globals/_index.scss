////
/// Globals entry point.
/// @group entry
/// @copyright IBM Security 2019 - 2020
////

@import 'theme/index';

@import 'variables/index';

@import 'carbon-components/scss/globals/scss/helper-mixins';

@import 'color/index';
@import 'grid/index';
@import 'layout/index';
@import 'motion/index';
@import 'type/index';

@import 'border/index';
@import 'deprecate/index';
@import 'feature-flags/index';
@import 'namespace/index';
@import 'rtl/index';
