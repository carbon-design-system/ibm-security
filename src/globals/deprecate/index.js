/**
 * @file Deprecation warning.
 * @copyright IBM Security 2019 - 2021
 */

import isDevelopment from '../env';

/**
 * Filters, compares, and logs the appropriate properties with a description.
 * @param {string} description The description to use when logging.
 * @param {string} type The property type to log.
 * @param {Array.<string>} componentKeys The component keys to filter.
 * @param {function} condition The condition to filter the component keys.
 * @returns {string} The description and appropriate properties.
 */
const logDeviations = (description, type, componentKeys, filterCondition) =>
  console.log(`${description} ${type}`, componentKeys.filter(filterCondition));

/**
 * Helper for comparing and logging deviations between multiple components.
 * @param {string} carbonComponent The Carbon component to compare.
 * @param {string} securityComponent The Security-specific component to compare.
 */
const compareDeviations = (carbonComponent, securityComponent) => {
  ['defaultProps', 'propTypes'].forEach(type => {
    const carbonComponentKeys = Object.keys(carbonComponent[type]);
    const securityComponentKeys = Object.keys(securityComponent[type]);

    logDeviations(
      'Common',
      type,
      carbonComponentKeys,
      key => securityComponentKeys.indexOf(key) !== -1
    );

    logDeviations(
      'Carbon',
      type,
      carbonComponentKeys,
      key => securityComponentKeys.indexOf(key) === -1
    );

    logDeviations(
      'Security',
      type,
      securityComponentKeys,
      key => carbonComponentKeys.indexOf(key) === -1
    );
  });
};

/**
 * Returns a standardized deprecation warning for development.
 * @param {string} warning - The warning value supplied.
 * @returns {console.trace} A stack trace containing the created deprecation details.
 */
function deprecationWarning(warning) {
  if (isDevelopment()) {
    try {
      throw new Error(
        `Warning: ${warning} and will be removed in an upcoming major release`
      );
    } catch (error) {
      console.warn(error);
    }
  }
}

/**
 * Returns a standardized deprecation warning.
 * @param {string} actual - The actual value supplied.
 * @param {string} expected - The expected value to be supplied.
 * @returns {console.trace} A stack trace containing the created deprecation details.
 */
const deprecate = (actual, expected) =>
  deprecationWarning(
    `'${actual}' has been deprecated in favor of '${expected}'`
  );

/**
 * Returns a standardized component deprecation warning.
 * @param {Object.<string, string>} deprecationWarning - An object containing properties that construct the deprecation warning.
 * @param {string} deprecationWarning.actual - The actual value supplied.
 * @param {string} deprecationWarning.componentName - The component name it was supplied to.
 * @param {string} deprecationWarning.description - The description of the deprecation warning.
 * @param {string} deprecationWarning.expected - The expected value to be supplied.
 * @returns {console.trace} A stack trace containing the created deprecation details.
 */
export default ({ actual, componentName, description, expected }) =>
  deprecationWarning(
    `Deprecated ${description} '${actual}' supplied to '${componentName}', expected '${expected}'`
  );

export { compareDeviations, deprecate };
