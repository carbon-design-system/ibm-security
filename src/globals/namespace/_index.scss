////
/// Assets related to namespacing selectors.
/// @group namespace
/// @copyright IBM Security 2019
////

@import '@carbon/import-once/scss/import-once';

/// The prefix used to namespace all selectors in the CSS.
/// @type String
/// @access private
$namespace: 'security--';

///
/// Appends the namespace for a component to another namespace.
/// @param {String} $component-namespace The component namespace to append to.
/// @param {String} $component-name The component namespace to append.
/// @return {string} The appended component namespace to return.
///
@function append-component-namespace($component-namespace, $component-name) {
  @return '#{$component-namespace}__#{$component-name}';
}

///
/// Creates the namespace for a component using the global namespace.
/// @param {String} $component-name The component namespace to create.
/// @return {String} The component namespace to return.
///
@function get-component-namespace($component-name) {
  @return '.#{$namespace}#{$component-name}';
}

/// Exports the content once to the unique namespace.
/// @param {String} $name The name to namespace.
/// @content The content to export once.
/// @example scss
///   @include export-namespace($name: platform) {
///     body {
///       background-color: $body__color__background;
///     }
///   }
/// @output css
///   body {
///     background-color: #fff;
///   }
///
@mixin export-namespace($name) {
  @include exports($name: #{$namespace}#{$name}) {
    @content;
  }
}

/// Sets the namespace prefix on to the given selector.
/// @param {String} $selector The selector string to namespace.
/// @content The styles for the namespaced component class name.
/// @example scss
///   @include namespace($selector: header) {
///     background-color: $header__color__background;
///   }
/// @output css
///   .security--header {
///     background-color: #fff;
///   }
///
@mixin namespace($selector) {
  #{get-component-namespace($component-name: $selector)} {
    @content;
  }
}
