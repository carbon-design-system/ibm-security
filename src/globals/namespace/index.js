/**
 * @file Namespace.
 * @copyright IBM Security 2018 - 2021
 */

const {
  settings: { prefix: carbonPrefix },
} = require('carbon-components');

const namespace = 'security--';

/**
 * Appends the namespace for a component to another namespace.
 * @param {string} componentNamespace The component namespace to append to.
 * @param {string} componentName The component namespace to append.
 * @returns {string} The appended component namespace to return.
 */
const appendComponentNamespace = (componentNamespace, componentName) =>
  `${componentNamespace}__${componentName}`;

/**
 * Creates the namespace for a component using the global namespace.
 * @param {string} componentName The component namespace to create.
 * @returns {string} The component namespace to return.
 */
const getComponentNamespace = componentName => `${namespace}${componentName}`;

export { appendComponentNamespace, carbonPrefix, getComponentNamespace };
