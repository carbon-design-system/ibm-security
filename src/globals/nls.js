/**
 * @file Default labels.
 * @copyright IBM Security 2019
 */

import PropTypes from 'prop-types';

export const labels = {
  // DataDecorator
  /** @type {string} the primary button label for the DataDecorator's PanelContainer */
  DATA_DECORATOR_PRIMARY_BUTTON: '',
  /** @type {string} the secondary button label for the DataDecorator's PanelContainer */
  DATA_DECORATOR_SECONDARY_BUTTON: '',
  /** @type {string} the close button aria label for the DataDecorator's PanelContainer */
  DATA_DECORATOR_CLOSE_BUTTON: 'Close',

  // FilterPanel
  /** @type {string} Label for truncated filters list to expand */
  FILTER_PANEL_CATEGORY_EXPAND_LABEL: '',
  /** @type {string} Label for expanded filters list to collapse */
  FILTER_PANEL_CATEGORY_COLLAPSE_LABEL: '',
  /** @type {string} Label for filter search input */
  FILTER_PANEL_SEARCH_LABEL: '',
  /** @type {string} Label for filter search when no filters are found */
  FILTER_PANEL_SEARCH_NO_RESULTS_LABEL: '',
  /** @type {string} The default aria label for the Search input clear button */
  FILTER_PANEL_SEARCH_CLOSE_BUTTON: 'Clear search',

  // Panel
  /** @type {string} The PanelContainer's primary button label */
  PANEL_CONTAINER_PRIMARY_BUTTON: '',
  /** @type {string} The PanelContainer's secondary button label */
  PANEL_CONTAINER_SECONDARY_BUTTON: '',
  /** @type {string} The PanelContainer's close button aria label */
  PANEL_CONTAINER_CLOSE_BUTTON: 'Close',

  // Search
  /** @type {string} aria label for the Search activate btn and input */
  SEARCH_LABEL: '',
  /** @type {string} Placeholder text to be displayed in the search input. */
  SEARCH_PLACEHOLDER_LABEL: '',
  /** @type {string} An aria label for the search input clear button. */
  SEARCH_CLOSE_BUTTON: 'Clear search',

  // `Tag`
  /** @type {string} Label for the button used to remove the tag. */
  TAG_REMOVE_BUTTON: 'Remove',

  // `TagWall`
  /** @type {string} Label. */
  TAG_WALL_LABEL: '',

  /** @type {string} Tag wall 'add' button. */
  TAG_WALL_ADD_BUTTON: '',

  /** @type {string} Tag 'remove' button. */
  TAG_WALL_REMOVE_BUTTON: '',

  // Tearsheet
  /** @type {string} label for the tearsheet primary button */
  TEARSHEET_PRIMARY_BUTTON: 'Okay',
  /** @type {string} label for the tearsheet secondary button */
  TEARSHEET_SECONDARY_BUTTON: 'Cancel',
  /** @type {string} label for the tearsheet delete button */
  TEARSHEET_DELETE_BUTTON: 'Delete',
  /** @type {string} aria label for the tearsheet close button */
  TEARSHEET_CLOSE_BUTTON: 'Close',
  /** @type {string} Text for the tearsheet cancel setup button */
  TEARSHEET_TERTIARY_BUTTON: 'Cancel setup',
  /** @type {string} Secondary text for the tearsheet cancel setup button */
  TEARSHEET_TERTIARY_SECONDARY_TEXT: 'Information will not be saved upon exit',

  // TearsheetSmall
  /** @type {string} label for the tearsheet primary button */
  TEARSHEET_SMALL_PRIMARY_BUTTON: 'Save',
  /** @type {string} label for the tearsheet secondary button */
  TEARSHEET_SMALL_SECONDARY_BUTTON: 'Cancel',
  /** @type {string} aria label for the tearsheet close button */
  TEARSHEET_SMALL_CLOSE_BUTTON: 'Close',

  // Wizard.
  /** @type {string} The finish button label for the Wizard component's Tearsheet */
  WIZARD_FINISH_BUTTON: 'Finish',
  /** @type {string} The next button label for the Wizard component's Tearsheet */
  WIZARD_NEXT_BUTTON: 'Next',
  /** @type {string} The cancel button label for the Wizard component's Tearsheet */
  WIZARD_CANCEL_BUTTON: 'Cancel',
  /** @type {string} The back button label for the Wizard component's Tearsheet */
  WIZARD_BACK_BUTTON: 'Back',
  /** @type {string} The cancel setup button for the Wizard component's Tearsheet */
  WIZARD_TERTIARY_BUTTON: 'Cancel',
  /** @type {string} Secondary text for the Wizard component Tearsheet cancel setup button */
  WIZARD_TERTIARY_SECONDARY_TEXT: 'Information will not be saved upon exit',
  /** @type {string} The delete button label for the Wizard component's Tearsheet */
  WIZARD_TEARSHEET_DELETE_BUTTON: 'Delete connection',

  // ErrorPage
  /** @type {object} List of Error objects */
  ERRORS: {
    404: {
      TITLE: '404',
      ERRORNAME: 'Page not found',
      ERRORMESSAGE:
        'We could not find the page you were looking for, but here are some helpful places to start from:',
    },
    403: {
      TITLE: '403',
      ERRORNAME: 'Access denied',
      ERRORMESSAGE:
        'You do not have the correct subscription to access this content. Please contact your administrator.',
    },
    500: {
      TITLE: '500',
      ERRORNAME: 'Internal server error',
      ERRORMESSAGE:
        'An internal server error occurred when trying to access this content.',
    },
    default: {
      TITLE: 'UnknownError',
    },
  },
};

// Helpers

/**
 * filters out falsy entries
 * @param {Object} obj object to be filtered
 * @returns {Object} object without falsey entries
 */
export const filterFalsey = obj =>
  Object.entries(obj)
    .filter(([key, value]) => Boolean(value)) // eslint-disable-line no-unused-vars
    .reduce((result, [key, value]) => {
      result[key] = value;
      return result;
    }, {});

export const propType = PropTypes.objectOf(
  PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object])
);
