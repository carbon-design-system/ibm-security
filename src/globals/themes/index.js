/**
 * @file Themes.
 * @copyright IBM Security 2019
 */

import { formatTokenName, g100 } from '@carbon/themes';

const theme = {};

// Rewrites the theme using CSS custom properties.
Object.keys(g100).forEach(token => {
  theme[token] = `var(--${formatTokenName(token)}, ${g100[token]})`;
});

export default theme;
