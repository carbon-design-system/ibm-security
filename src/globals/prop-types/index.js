/**
 * @file Custom `prop-types` validation.
 * @copyright IBM Security 2018
 */

import { checkPropTypes } from 'prop-types';
import deprecate from '../deprecate';

/**
 * Throws a warning for deprecated props and validates that prop.
 * @param {string} replacementPropName The suggested replacement prop name.
 * @param {PropTypes.<*>} propValidation The prop validation to check against.
 */
const deprecatedProp =
  (replacementPropName, propValidation) =>
  (props, deprecatedPropName, componentName) => {
    /**
     * Creates a prop object to validate against.
     * @param {string|PropTypes.<*>} propValue The prop value to add to the object.
     * @returns {object.<string, *>} The prop object to validate against.
     */
    const createPropObject = (propValue) => ({
      [`${deprecatedPropName}`]: propValue,
    });

    const prop = props[deprecatedPropName];

    if (prop) {
      deprecate({
        actual: deprecatedPropName,
        componentName,
        description: 'prop type',
        expected: replacementPropName,
      });
    }

    checkPropTypes(
      createPropObject(propValidation),
      createPropObject(prop),
      deprecatedPropName,
      componentName
    );
  };

export default deprecatedProp;
