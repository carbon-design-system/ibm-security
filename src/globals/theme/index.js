/**
 * @file Themes.
 * @copyright IBM Security 2019 - 2021
 */

import { formatTokenName, g100 } from '@carbon/themes';

const theme = {};

// Rewrites the theme using CSS custom properties.
Object.keys(g100).forEach(token => {
  theme[token] = `var(--cds-${formatTokenName(token)}, ${g100[token]})`;
});

export default theme;
