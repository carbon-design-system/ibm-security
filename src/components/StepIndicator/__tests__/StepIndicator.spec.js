/**
 * @file Step indicator tests.
 * @copyright IBM Security 2019 - 2021
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderWithinLandmark from '../../../../config/jest/helpers/renderWithinLandmark';

import { Step, StepIndicator } from '../../..';

import { carbonPrefix } from '../../../globals/namespace';

describe('StepIndicator', () => {
  test('should have no Axe or DAP violations', async () => {
    const { container } = renderWithinLandmark(
      <StepIndicator currentIndex={1}>
        <Step label="test label 1" description="test description 1" />
        <Step label="test label 2" description="test description 2" />
      </StepIndicator>
    );
    await expect(container).toHaveNoAxeViolations();
    await expect(container).toHaveNoDAPViolations('StepIndicator');
  });

  test('should add a custom class to the parent component', () => {
    render(<StepIndicator className="custom-class" />);
    expect(document.querySelector('.custom-class')).toBeVisible();
  });

  test('should pass through extra prop to parent component via spread attriute', () => {
    const { queryByTestId } = render(<StepIndicator data-testid="test-id" />);
    expect(queryByTestId('test-id')).toBeVisible();
  });

  test('should render even when `currentIndex` is `null`', () => {
    const { queryByText } = render(
      <StepIndicator>
        <Step label="test label" />
      </StepIndicator>
    );
    expect(queryByText(/test label/i)).toBeVisible();
  });

  test('should apply `current={true}` to the child identified with `currentIndex` value', () => {
    const { queryByText } = render(
      <StepIndicator currentIndex={1}>
        <Step label="test label 1" />
        <Step label="test label 2" />
      </StepIndicator>
    );
    // A step with `current={true}` will recieve a specific class:
    expect(queryByText('test label 2').closest('li')).toHaveClass(
      `${carbonPrefix}--progress-step--current`
    );
  });
});
