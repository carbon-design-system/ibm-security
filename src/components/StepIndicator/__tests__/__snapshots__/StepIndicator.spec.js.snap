// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepIndicator render renders correctly 1`] = `
<ProgressIndicator
  className="security--step-indicator"
  currentIndex={1}
>
  <Step
    className=""
    complete={null}
    current={null}
    description="Step 1: Getting started with Carbon Design System"
    disabled={null}
    invalid={null}
    label="First step"
  />
  <Step
    className=""
    complete={null}
    current={null}
    description="Step 2: Getting started with Carbon Design System"
    disabled={null}
    invalid={null}
    label="Second step"
  />
</ProgressIndicator>
`;

exports[`StepIndicator render renders if \`currentIndex\` is null 1`] = `
<ul
  class="bx--progress security--step-indicator"
>
  <li
    class="bx--progress-step bx--progress-step--incomplete security--step"
  >
    <button
      class="bx--progress-step-button bx--progress-step-button--unclickable"
      tabindex="-1"
    >
      <span
        class="bx--assistive-text"
      >
        Incomplete
      </span>
      <svg
        aria-hidden="true"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        style="will-change:transform"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16,2A14,14,0,1,0,30,16,14,14,0,0,0,16,2Zm0,26A12,12,0,1,1,28,16,12,12,0,0,1,16,28Z"
        />
        <title>
          Step 2: Getting started with Carbon Design System
        </title>
      </svg>
      <p
        class="bx--progress-label"
      >
        Second step
      </p>
      <span
        class="bx--progress-line"
      />
    </button>
  </li>
</ul>
`;

exports[`StepIndicator render renders the HTML of the node's subtree 1`] = `
<ul
  class="bx--progress security--step-indicator"
>
  <li
    class="bx--progress-step bx--progress-step--complete security--step"
  >
    <button
      class="bx--progress-step-button bx--progress-step-button--unclickable"
      tabindex="-1"
    >
      <span
        class="bx--assistive-text"
      >
        Complete
      </span>
      <svg
        aria-hidden="true"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        style="will-change:transform"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16,2A14,14,0,1,0,30,16,14,14,0,0,0,16,2Zm0,26A12,12,0,1,1,28,16,12,12,0,0,1,16,28Z"
        />
        <path
          d="M14 21.5L9 16.54 10.59 14.97 14 18.35 21.41 11 23 12.58 14 21.5z"
        />
        <title>
          Step 1: Getting started with Carbon Design System
        </title>
      </svg>
      <p
        class="bx--progress-label"
      >
        First step
      </p>
      <span
        class="bx--progress-line"
      />
    </button>
  </li>
  <li
    class="bx--progress-step bx--progress-step--current security--step"
  >
    <button
      class="bx--progress-step-button bx--progress-step-button--unclickable"
      tabindex="-1"
    >
      <span
        class="bx--assistive-text"
      >
        Current
      </span>
      <svg
        aria-hidden="true"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        style="will-change:transform"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16,2A14,14,0,1,0,30,16,14,14,0,0,0,16,2Zm0,26A12,12,0,1,1,28,16,12,12,0,0,1,16,28Z"
        />
        <path
          d="M16,10a6,6,0,1,0,6,6A6,6,0,0,0,16,10Z"
        />
        <title>
          Step 2: Getting started with Carbon Design System
        </title>
      </svg>
      <p
        class="bx--progress-label"
      >
        Second step
      </p>
      <span
        class="bx--progress-line"
      />
    </button>
  </li>
</ul>
`;
